### PR TITLE
added data saver class and integrated with simulator

### PR DESCRIPTION
**Pull request by cryingpotat0**
Created a data saver class that's more generic and allows us to save arbitrary data for arbitrary cars. Might be overkill for this use case? I felt that it was the cleanest way to write without adding a bunch of functions/ arguments to the simulator. I could change it to be just more functions in sim.py if needed, but I felt that the number of arguments for the constructor would be too bloated. 

Another option is to abstract away the data saver class and create it using a function within sim.py, which would allow us to have fewer external class dependencies for the user as well.

### CODE REVIEW

**Comment by cryingpotat0** [comment]
The data is incredibly dense, we might want to consider compressing it somehow? For reference it was about 750 MB for a 100 timesteps with the default number of cars and pedestrians using only the birdseye observation + steeringAcc action. We could look into using np.save or other hd5f storage options.

File: fluids/datasaver.py
Comment (jerryz123) [inline fluids/datasaver.py]: I'm surprised this produces dumps as large as you say they are. `np.save` is a good alternative if pickling does not work out.

**Review by jerryz123 — CHANGES_REQUESTED** [review]

File: fluids/datasaver.py
Comment (cryingpotat0) [reply fluids/datasaver.py]: np.save is giving me essentially identical performance, although I suspect there is not a lot of compression happening. This is because when I compress the 750mb file to a zip, I get a 6mb file back out. We could try something like this (https://stackoverflow.com/questions/22400652/compress-numpy-arrays-efficiently) directly?

Comment (jerryz123) [reply fluids/datasaver.py]: Hmm what are the parameters of the observation? Like type, resolution, etc. I want to calculate if 750 MB is reasonable for what we are collecting.

Compressing it is a good idea. I think our data is pretty compressible.

Comment (cryingpotat0) [reply fluids/datasaver.py]: Here's the breakdown for how much data we should be getting. I'm only considering the observation because it's way larger than the action space. The test I did was for the birds_eye_observation which consister of (500,500,3) array of uint8s. 
This means the total data should be 500x 500x 3 x 8bits x 10cars x 100 timesteps which is exactly 750 megabytes, so numpy does not comrpress while saving.

Comment (cryingpotat0) [reply fluids/datasaver.py]: .npz is 6mb for 100 timesteps, pushing the new code now. Using numpy as a storage mechanism does make the data storage a bit more complicated because I have to define a custom data type for the numpy array, but I still think it's better than the pickle based approach because it seems cleaner to interface with.

Comment (cryingpotat0) [reply fluids/datasaver.py]: One downside is that it takes about 30 seconds to compress the file each time, so we should decide what batch size to dump data at to maximize speed. I can also test with other compression schemes sometime tomorrow, so it's probably best to hold off merging the pull request until then

File: fluids/datasaver.py
Comment (cryingpotat0) [inline fluids/datasaver.py]: Need to modify description once finalized

File: fluids/datasaver.py
Comment (jerryz123) [reply fluids/datasaver.py]: Sounds good, thanks for doing this analysis. 

I think we will probably usually be dumping the GridObservations anyways, not the BirdsEye, so maybe that will decrease the data rate significantly.

**Comment by cryingpotat0** [comment]
savez_compressed: Saved compressed file in 37.13291621208191 msecs
gzip: Saved compressed file in 50.064592123031616s

**Comment by cryingpotat0** [comment]
Pushed final version, can be merged once travis completes

**Review by jerryz123 — APPROVED** [review]

**Pull request merged**
